
These are questions from previous exams, I would really appreciate you send me potential questions for the next exam

- What is NGS useful for?
- Why is NGS so popular? What is your opinion in this respect, do you believe we do or we do not need so much sequence data?
- What do we mean by saying NGS data are 'noisy'? Which problems can we encounter because of that?
- What advantages does sequencing RNA (RNAseq) offer vs expression arrays?
- What is the purpose of NGS data alignment? Why don't we use classical tools such as clustal?
- Are there usually errors in reference genomes? If so, what can be their impact?
- If there is no reference genome for the species of our interest, what can we do? 
- Can you name some public repositories for NGS data?
- In what circumstances would Sanger sequencing be preferred to NGS?
- In what circumstances can high density genotyping, say using a commercial array, be preferred to NGS? 
- Why is it important to filter by quality in any NGS bioinformatics step?
- What is the advantage of using Phred scores for coding base quality?
- Name some differences between NGS and Sanger sequencing.
- Who was Fred Sanger? Why is he famous for?
- What are the main differences between 'second' and 'third' NGS technologies?
- Why are structural variants difficult to detect with NGS?
- What is a VCF file refered to when, in a sample, we observe 0/0, 0/1 and 1/1 in the genotype columns?
- Do you know the difference between 0/1 and 0|1 notation in the genotype column of a vcf file?
- Suppose you have money to sequence 100 individuals at 1x or 1 individual at 100x, which one would you choose? would any of these extremes be useful and what for in case they are useful?
- Ancient DNA can be sequenced, but do you know which are the main problems we face?
- What are the benefits of using multiple sample SNP calling instead of single sample SNP calling?
- Why is it needed to use a minimum and maximum read depth to ensure an adequate NGS Data Anaysis?

